### ISSUE TITLE
System.NullReferenceException: 'Object reference not set to an instance of an object

### ISSUE THREAD

**Issue by imdshalik** [Job79/EasyTcp#88]
EasyTcp.Client.EasyTcpClient.Socket.get returned null. 

Getting null value after every connection

**Comment by Job79** [comment]
Could you give some example code to reproduce this issue? The unit tests seem to work fine, and they are making plenty of new connections.

**Comment by imdshalik** [comment]
![Exception Error](https://user-images.githubusercontent.com/108178134/209278748-bf35232e-19bf-4051-820e-0111095fd91e.png)
Client = new EasyTcpClient();
           
            Client.DataReceived += (object sender, Message e) =>
            {
                try
                {
                    firsttime = 2;
                    ReTryCountwithdisclose = 0;
                  
                    byte[] msgHeaderBytes = e.Data;
                    string msgHeaderHex = BufferManipulator.ToHexString(msgHeaderBytes, false);

                    msgHeaderHex = msgHeaderHex.StartsWith("02") ? "0100" + msgHeaderHex : msgHeaderHex;

                    logger.WithProperty("MethodName", "EasyTcpClient.DataReceived").Info($"msgHeader - {msgHeaderHex}");

                    EasyTcpClient _cl = (EasyTcpClient)sender;
                    
                    Thread.Sleep(2000);

                    //if(Convert.ToString(_cl.Socket.Available)!="")
                    if (!String.IsNullOrEmpty(Convert.ToString(_cl.Socket.Available)))
                        {
                        rbytes = _cl.Socket.Available;
                        }
                        else
                        {
                           //firsttime = 1;
                           rbytes = 0;
                        }

This is the code snippet after first time c1.socket - sends null, and exception i have mentioned

**Comment by Job79** [comment]
I think the following is happening here:
1. Client receives a message
2. Client starts 2 second sleep
3. Client disconnects (this sets the socket to null)
4. Client continues from sleep and crashes

I don't think the library can really help here. You have to check whether the socket is null, or beter: use the library function `IsConnected` on the EasyTcpClient.

**Comment by imdshalik** [comment]
Yeah socket is going to null state but in my code i am not disconnecting or closing i dont know why it going to null state after every connection. Will updating the package will work?

**Comment by mzelecom** [comment]
i have the same Issue.
After some time, the server keeps giving "Object reference not set to an instance of an object" error. In client.send().

**Comment by imdshalik** [comment]
Will updating the package will work?
Please Reply

**Comment by Job79** [comment]
Setting the socket to null is probably not the best behavior, and this could be fixed with a package update but I don't think this would solve your problems. It would only change the error that occurs. Sending a message to a client that disconnected simply isn't possible.

Do you know why the client disconnects? While very unlikely there might be another problem that causes the client to disconnect. If this is the case, could you send an example so I can reproduce the problem?

It might be that I am misunderstanding the problem, but to me it seems like you are trying to send data to a disconnected client. This will always be a problem, and can only be fixed by using a try catch. Checking whether the socket is null or the connection is active would not work, because the client might disconnect right after checking the connection but before sending a message.

**Comment by imdshalik** [comment]
Dear 
I am using shutdown method to close the connection like this 
![image](https://user-images.githubusercontent.com/108178134/209901116-5d726a1c-f67c-45f8-a19a-24f9869b9983.png)
instead will the below things work
Client.Socket.Close();
Client.Disconnect();

I am using Client.Socket.Shutdown(SocketShutdown.Both); to disconnect suppose if i wont use this first time i can send and receive the data properly after that I can't receive any i.e No transaction is happening.

I will update the package and will check also i will remove all disconnect statement in code then will check.

I am using in POS operation, previously I kept the connections open but that time i am facing problem after the first transaction i cant do any thing in the receiver side i am not receiving any. So ly i use shutdown method to close connection after the transaction

**Comment by mzelecom** [comment]
i never disconnected client.
after 4 hour and 1500 client connected to server .it will show error on all client.

**Comment by Job79** [comment]
After 4 hours the complete server crashes? Did the process exit?
Sounds like a issue, but I will need more information. Did you get a stack trace? What happend before they disconnected?
